Add ``onnx`` backend to ``torch.export`` API

thiagocrepaldi opened the pull request:
Related to https://github.com/pytorch/pytorch/issues/109131

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109650
* #109649

This PR adds the ONNX exporter backend for ``torch.export``.
The ONNX exporter will leverage Torch Dynamo under the hood, but produce
a ONNX model as a result of the export. The ONNX backend will also
enable ONNX Runtime to transparently execute the ONNX model.

An example usage is shown below:

```python
import torch

def func(x):
    y = x + 1
    z = y.relu()
    return (y, z)
inputs = (torch.randn(1, 1, 2),)
export_options = torch.onnx.ExportOptions()
exported_onnx = torch.export.export(func, inputs, backend="onnx", options={"export_options":  export_options})

onnxruntime_output = exported_onnx(*inputs)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @Xia-Weiwen

Commit:
Add ``onnx`` backend to ``torch.export`` API

This PR adds the ONNX exporter backend for ``torch.export``.
The ONNX exporter will leverage Torch Dynamo under the hood, but produce
a ONNX model as a result of the export. The ONNX backend will also
enable ONNX Runtime to transparently execute the ONNX model.

An example usage is shown below:

```python
import torch

def func(x):
    y = x + 1
    z = y.relu()
    return (y, z)
inputs = (torch.randn(1, 1, 2),)
export_options = torch.onnx.ExportOptions()
exported_onnx = torch.export.export(func, inputs, backend="onnx", options={"export_options":  export_options})

onnxruntime_output = exported_onnx(*inputs)
```

[ghstack-poisoned]

Commit:
Update on "Add ``onnx`` backend to ``torch.export`` API"

Related to https://github.com/pytorch/pytorch/issues/109131


This PR adds the ONNX exporter backend for ``torch.export``.
The ONNX exporter will leverage Torch Dynamo under the hood, but produce
a ONNX model as a result of the export. The ONNX backend will also
enable ONNX Runtime to transparently execute the ONNX model.

An example usage is shown below:

```python
import torch

def func(x):
    y = x + 1
    z = y.relu()
    return (y, z)
inputs = (torch.randn(1, 1, 2),)
export_options = torch.onnx.ExportOptions()
exported_onnx = torch.export.export(func, inputs, backend="onnx", options={"export_options":  export_options})

onnxruntime_output = exported_onnx(*inputs)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -612,6 +612,38 @@ def __init__(
+        """
+        import onnxruntime  # type: ignore[import]
+
+        onnx_input = self.adapt_torch_inputs_to_onnx(*args, **kwargs)
```
Nice to see adapter in action.

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -612,6 +612,38 @@ def __init__(
+            k.name: to_numpy(v) for k, v in zip(ort_session.get_inputs(), onnx_input)
+        }
+
+        return ort_session.run(None, onnxruntime_input)
```
It will be good to have an `adapt_onnx_output_to_torch`, which does not exist yet. Doesn't have to be this PR.

BowenBao reviewed (commented):
Are there other changes to `ExportOutput` other than renaming and a new `__call__` method?

Commit:
Update on "Add ``onnx`` backend to ``torch.export`` API"

Related to https://github.com/pytorch/pytorch/issues/109131


This PR adds the ONNX exporter backend for ``torch.export``.
The ONNX exporter will leverage Torch Dynamo under the hood, but produce
a ONNX model as a result of the export. The ONNX backend will also
enable ONNX Runtime to transparently execute the ONNX model.

An example usage is shown below:

```python
import torch

def func(x):
    y = x + 1
    z = y.relu()
    return (y, z)
inputs = (torch.randn(1, 1, 2),)
export_options = torch.onnx.ExportOptions()
exported_onnx = torch.export.export(func, inputs, backend="onnx", options={"export_options":  export_options})

onnxruntime_output = exported_onnx(*inputs)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

Commit:
Update on "Add ``onnx`` backend to ``torch.export`` API"

Related to https://github.com/pytorch/pytorch/issues/109131


This PR adds the ONNX exporter backend for ``torch.export``.
The ONNX exporter will leverage Torch Dynamo under the hood, but produce
a ONNX model as a result of the export. The ONNX backend will also
enable ONNX Runtime to transparently execute the ONNX model.

An example usage is shown below:

```python
import torch

def func(x):
    y = x + 1
    z = y.relu()
    return (y, z)
inputs = (torch.randn(1, 1, 2),)
export_options = torch.onnx.ExportOptions()
exported_onnx = torch.export.export(func, inputs, backend="onnx", options={"export_options":  export_options})

onnxruntime_output = exported_onnx(*inputs)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

Commit:
Update on "Add ``onnx`` backend to ``torch.export`` API"

Related to https://github.com/pytorch/pytorch/issues/109131


This PR adds the ONNX exporter backend for ``torch.export``.
The ONNX exporter will leverage Torch Dynamo under the hood, but produce
a ONNX model as a result of the export. The ONNX backend will also
enable ONNX Runtime to transparently execute the ONNX model.

An example usage is shown below:

```python
import torch

def func(x):
    y = x + 1
    z = y.relu()
    return (y, z)
inputs = (torch.randn(1, 1, 2),)
export_options = torch.onnx.ExportOptions()
exported_onnx = torch.export.export(func, inputs, backend="onnx", options={"export_options":  export_options})

onnxruntime_output = exported_onnx(*inputs)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

Commit:
Update on "Add ``onnx`` backend to ``torch.export`` API"

Related to https://github.com/pytorch/pytorch/issues/109131


This PR adds the ONNX exporter backend for ``torch.export``.
The ONNX exporter will leverage Torch Dynamo under the hood, but produce
a ONNX model as a result of the export. The ONNX backend will also
enable ONNX Runtime to transparently execute the ONNX model.

An example usage is shown below:

```python
import torch

def func(x):
    y = x + 1
    z = y.relu()
    return (y, z)
inputs = (torch.randn(1, 1, 2),)
export_options = torch.onnx.ExportOptions()
exported_onnx = torch.export.export(func, inputs, backend="onnx", options={"export_options":  export_options})

onnxruntime_output = exported_onnx(*inputs)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

thiagocrepaldi commented:
Closing this for now
